src/vmspawn/vmspawn-util.c

```diff
@@ -0,0 +1,195 @@
+                        return 0;
+                }
+        }
+
+#undef FIRMWARE_LOCATIONS
+#ifdef __x86_64__
+        #define FIRMWARE_LOCATIONS "/usr/share/ovmf/ovmf_code_x64.bin", "/usr/share/ovmf/x64/OVMF_CODE.fd", "/usr/share/qemu/ovmf-x86_64.bin"
+#elif __i386__
+        #define FIRMWARE_LOCATIONS "/usr/share/ovmf/ovmf_code_ia32.bin", "/usr/share/edk2/ovmf-ia32/OVMF_CODE.fd"
+#elif __aarch64__
+        #define FIRMWARE_LOCATIONS "/usr/share/AAVMF/AAVMF_CODE.fd"
+#elif __arm__
+        #define FIRMWARE_LOCATIONS "/usr/share/AAVMF/AAVMF32_CODE.fd"
+#else
+        #define FIRMWARE_LOCATIONS
+#endif
+
+        FOREACH_STRING(firmware, FIRMWARE_LOCATIONS) {
+                _cleanup_close_ int fd = -EBADF;
+                fd = open(firmware, O_CLOEXEC | O_RDONLY);
+                if (fd >= 0) {
+                        log_warning("Couldn't find OVMF firmware blob with secure boot support, "
+                                    "falling back to OVMF firmware blobs without secure boot support.");
+
+                        if (ret_firmware_path)
+                                *ret_firmware_path = firmware;
+                        return 1;
+                }
+        }
+
+#undef FIRMWARE_LOCATIONS
+        #define FIRMWARE_LOCATIONS \
+                "/usr/share/edk2/ovmf/OVMF_CODE.secboot.fd", \
+                "/usr/share/edk2-ovmf/OVMF_CODE.secboot.fd", \
+                "/usr/share/qemu/OVMF_CODE.secboot.fd", \
+                "/usr/share/ovmf/OVMF.secboot.fd", \
+                "/usr/share/OVMF/OVMF_CODE.secboot.fd"
+
+        FOREACH_STRING(firmware, FIRMWARE_LOCATIONS) {
+                _cleanup_close_ int fd = -EBADF;
+                fd = open(firmware, O_CLOEXEC | O_RDONLY);
+                if (fd >= 0) {
+                        if (ret_firmware_path)
+                                *ret_firmware_path = firmware;
+                        return 0;
+                }
+        }
+
+#undef FIRMWARE_LOCATIONS
+        #define FIRMWARE_LOCATIONS \
+                "/usr/share/edk2/ovmf/OVMF_CODE.fd", \
+                "/usr/share/edk2-ovmf/OVMF_CODE.fd", \
+                "/usr/share/qemu/OVMF_CODE.fd", \
+                "/usr/share/ovmf/OVMF.fd", \
+                "/usr/share/OVMF/OVMF_CODE.fd"
+
+        FOREACH_STRING(firmware, FIRMWARE_LOCATIONS) {
+                _cleanup_close_ int fd = -EBADF;
+                fd = open(firmware, O_CLOEXEC | O_RDONLY);
+                if (fd >= 0) {
+                        log_warning("Couldn't find OVMF firmware blob with secure boot support, "
+                                    "falling back to OVMF firmware blobs without secure boot support.");
+
+                        if (ret_firmware_path)
+                                *ret_firmware_path = firmware;
+                        return 1;
+                }
+        }
+
+#undef FIRMWARE_LOCATIONS
+        if (ret_firmware_path)
+                *ret_firmware_path = NULL;
```

Comment:
in systemd we have this rule that return parameters should be initialized in all cases of success and never in cases of failure. Here you return `-ENOENT`, i.e. an error, hence the parameter should not be touched.

Reply:
Fixed in 111681abaf.